### ISSUE TITLE
Undefined is not a function - client.js: 498

### ISSUE THREAD

**Issue by MaffooBristol** [vote539/socketio-file-upload#19]
Appears sporadically when I'm trying to upload images, haven't been able to perfectly recreate the environment that breaks it, but it fails on this line:

```javascript
// CONSTRUCTOR: Listen to the "complete", "ready", and "error" messages on the socket.
socket.on("siofu_ready", function(data){
	readyCallbacks[data.id](data.name);
});
```

... it makes me think that for some reason `data.id` is sometimes undefined. I've used the api in the same way as in the documentation so I don't know what's going on.

Cheers, Matt.

**Comment by vote539** [comment]
Hi Matt,

Could you look in the Network tab at the Web Inspector and try and see what is the value of "id" when it is sent on the outgoing "siofu_start" event and the incoming "siofu_ready" event?  You can find the WebSocket that Socket.IO opens, and when you click on it there should be an option called "Frames" where you can inspect the packets going in and out.

-shane

**Comment by MaffooBristol** [comment]
Hey shane, thanks for replying!

Okay, well the results from what you suggested are pretty much the same as I got from console.logging out `readyCallbacks` and `data`, and the problem actually seems to be to do with event listeners maybe being called more than once. I've realised that I can reproduce the error actually: the web application I'm working on uses backbone/marionette routes; when I first go to the page and run the following SIOFU code, it works fine each time, however if I navigate away from the page and then back again (without a page reload) it seems to fire the events multiple times.

It also generates a write after end error on the server side to `uploader.on("error")` on subsequent uploads as well as increasing amounts of calls to `.on("saved")`.

First page load and image upload:
![image](https://cloud.githubusercontent.com/assets/3296828/5905620/2eb07fb4-a589-11e4-9d42-bc53f396ab68.png)

Second time the code is executed:
![image](https://cloud.githubusercontent.com/assets/3296828/5905606/212d5e7a-a589-11e4-83b1-dad84bb53051.png)

Third time, etc:
![image](https://cloud.githubusercontent.com/assets/3296828/5905633/4d718ab0-a589-11e4-9513-b7c3ea8eea2d.png)

And then the code, which should only bind the events once as each time it's run a new SocketIOFileUpload instance is created. Any ideas? Thanks :smile:

```coffeescript
app = require 'application'

module.exports = class AddNewAbstract extends Backbone.Marionette.ItemView

  templateHelpers: {}
  tempID: ''
  uploader: {}

  initialize: () ->
    @templateHelpers.title = @options.route.title
    for i in [0...6]
      @tempID += Math.floor((1 + Math.random()) * 0x10000).toString(16)

  bindImageUpload: (opts = {}) ->

    if !opts.uploadInput? or !opts.uploadDiv? or !opts.type?
      return console.error "Missing options"

    if !app.socket
      return console.error "Couldn't bind to socket."

    if !document.getElementById opts.uploadInput then return
    if !document.getElementById opts.uploadDiv then return

    @uploader = new SocketIOFileUpload app.socket

    @uploader.listenOnInput document.getElementById opts.uploadInput

    document.getElementById opts.uploadDiv
      .addEventListener 'click', @uploader.prompt, false

    @uploader.addEventListener 'start', (event) =>
      console.log 'test'
      event.file.meta =
        type:    opts.type
        private: true
        tempID:  @tempID

    app.socket.once 'file:saved', (filename) =>
      randKey = Math.floor((1 + Math.random()) * 0x10000).toString(16)
      $("##{opts.uploadDiv}").css
        backgroundImage: "url('#{filename}?#{randKey}')"
```

**Comment by vote539** [comment]
Okay, that's interesting.  I think what's happening is that the "old" uploader instances aren't destroyed, so when the user creates a new "AddNewAbstract" view, then you have two uploader instances listening at the same time.  The IDs increment starting at 0, independently for each instance.  This means that if you've already uploaded an image from the first time you had your "AddNewAbstract" view, the second time you make that view and upload an image, the first uploader will assign it an ID of 1 and the second image will assign it an ID of 0.  But when the result comes through, the new uploader still receives the event for file ID 1, which it can't find.

To fix your issue, I would make sure that you destroy old uploader instances when your "AddNewAbstract" view is destroyed.  It's been a while since I've worked on a project that uses Backbone/Marionette but I'm pretty sure there's some sort of "destroy" handler that you can bind to.  Read [here](https://github.com/vote539/socketio-file-upload#instancedestroy) for instructions on how to destroy an existing uploader instance.

If it's not possible to bind an event to the "AddNewAbstract" destroy event, what you could do would be to make a "global" uploader instance.  You would still call `uploader.listenOnInput document.getElementById opts.uploadInput` and `document.getElementById(opts.uploadDiv).addEventListener 'click', uploader.prompt, false`, only you wouldn't be creating an all-new uploader instance each time.

**Comment by MaffooBristol** [comment]
Ah yes, that makes sense. I'll have a play with destroying the instances on view destruction, it sounds in theory like it should work. Failing that I'll try the global instance suggestion- it may be more manageable as far as the architecture of the application goes anyway. Thanks!

**Comment by MaffooBristol** [comment]
I tried both methods, and both methods worked! Bonzer! :+1:

I noticed that when I use the first method:

```coffeescript
  onDestroy: () ->
    document.getElementById @opts.uploadDiv
      .removeEventListener 'click', @uploader.prompt, false
    @uploader.destroy()
    @uploader = null
```
... the errors are no longer there, but the `"siofu_ready"` callback still gets fired the respective amount of times that the code is executed, although the `readyCallbacks` array never expands past one value- however when I define SIOFU globally (or at least in the same scope as my socket), it builds up the `readyCallbacks` array each time an image is uploaded, which feels more like the intended effect. I can't imagine that it would cause any sort of memory issues would it, if the user had their session open for a long time and had uploaded lots and lots of files? Is there any form of "garbage collection", where the old callbacks that are long dead and not needed can be cleared from memory?

Either way, those questions are quite minor, it seems to be working bug free now. Thanks again mate!
